spec/app/ics-101-interchain-swap/README.md

```diff
@@ -0,0 +1,1613 @@
+                 amount: 0,
+                 denom: denom[i],
+               },
+               weight: weights[i],
+               decimal: decimals[i],
+            })
+        }
+    }
+  }
+
+  function findDenomBySide(side: PoolAssetSide): string | undefined {
+    for (const asset of this.pool.assets) {
+        if (asset.side === side) {
+            return asset.balance.denom;
+        }
+    }
+    return;
+}
+
+function findAssetBySide(side: PoolAssetSide): Coin | undefined {
+    for (const asset of this.pool.assets) {
+        if (asset.side === side) {
+            return asset.balance;
+        }
+    }
+    throw undefined;
+}
+
+function findPoolAssetBySide(side: PoolAssetSide): PoolAsset | undefined {
+    for (const asset of this.pool.assets) {
+        if (asset.side === side) {
+            return asset;
+        }
+    }
+    return;
+}
+
+function updateAssetPoolSide(denom: string, side: PoolAssetSide): PoolAsset | undefined {
+    for (const asset of this.pool.assets) {
+        if (asset.balance.denom === denom) {
+            asset.side = side;
+            return asset;
+        }
+    }
+    return undefined;
+}
+
+function addAsset(token: Coin): void {
+    for (const asset of this.pool.assets) {
+        if (asset.balance.denom === token.denom) {
+            asset.balance.amount += token.amount;
+            return;
+        }
+    }
+}
+
+function subtractAsset(token: Coin): Coin | undefined {
+    for (const asset of this.pool.assets) {
+        if (asset.balance.denom === token.denom) {
+            asset.balance.amount -=token.amount;
+            return asset.balance;
+        }
+    }
+    return
+}
+
+function addPoolSupply(token: Coin): void {
+    if (token.denom !== this.pool.id) {
+        return
+    }
+    this.supply.amount += token.amount;
+}
+
+function subtractPoolSupply(token: Coin): void {
+    if (token.denom !== this.pool.id) {
+        return
+    }
+    ilp.supply.amount -= token.amount;
+}
+}
+```
+
+```ts
+function generatePoolId(sourceChainId: string, destinationChainId: string, denoms: string[]): string {
+  const chainPrefix = [sourceChainId, destinationChainId].sort().join(",");
+  const id = chainPrefx.concat(denoms.sort().join(","));
+  const poolId = "pool" + sha256.hash(id);
+  return poolId;
+}
+```
+
+#### Interchain Market Maker
+
+The `InterchainMarketMaker` is a core component for swap calculations. It is initialized with an `InterchainLiquidityPool` and `feeRate`. Since the `feeRate` for a trading pair can be updated through governance, it is possible to have different rates on each chain.
+
+```ts
+class InterchainMarketMaker {
+    pool :InterchainLiquidityPool
+    static initialize(pool: InterchainLiquidityPool) : InterchainMarketMaker {
+        return {
+            pool: pool
+        }
+    }
+
+    // MarketPrice Bi / Wi / (Bo / Wo)
+    function marketPrice(denomIn, denomOut string): float64 {
+        const tokenIn = this.Pool.findAssetByDenom(denomIn)
```

Comment:
I think the `findAssetByDenom` function definition is missing from `InterchainLiquidityPool` interface.